proposals/p2875.md

```diff
@@ -0,0 +1,514 @@
+implementing an interface.
+
+```
+class X { var n: i32; }
+// ✅ OK, but inadvisable.
+impl {.a: X} as Call(()) where .Result = i32 {
+  fn Op[self: Self](args: ()) -> i32 {
+    return self.a.n;
+  }
+}
+fn Run() -> i32 {
+  // Returns 1.
+  return {.a = {.n = 1} as X}();
+}
+```
+
+There is no way to directly define a function-like callable class type that
+takes a compile-time argument. This is similar to the situation in C++ where
+there is no way to define an `operator()` for a value `x` of class type so that
+`x<T>()` passes `T` to `operator()` at compile time. However, this can be worked
+around by putting the compile-time value in the type of an argument instead:
+
+```
+class Wrap(T:! type) {}
+class Callable {
+  impl forall [T:! Printable] as Call((Wrap(T), T)) where .Result = () {
+    fn Op[self: Self](args: (Wrap(T), T)) {
+      let (_: auto, v: T) = args;
+      Print(v);
+    }
+  }
+}
+fn CallIt[F:! Call(Wrap(i32), i32)](f: F) {
+  f({} as Wrap(i32), 0);
+}
+fn Run() {
+  CallIt({} as Callable);
+}
+```
+
+## Rationale
+
+Principles:
+
+-   [Principle: one static open extension mechanism](/docs/project/principles/static_open_extension.md).
+    -   Overloaded calls are supported by implementing an interface.
+-   [Principle: Prefer providing only one way to do a given thing](/docs/project/principles/one_way.md).
+    -   There is only one obvious way to pass a callable to a generic function,
+        and it works efficiently whether the argument is a function, a callable
+        object, or (eventually) a lambda.
+
+Goals:
+
+-   [Performance-critical software](/docs/project/goals.md#performance-critical-software)
+    -   Passing around functions is no less efficient than passing around
+        callable objects.
+-   [Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)
+    -   Function types in C and C++ are notoriously hard to read. We avoid this
+        problem by not having signature-based function types.
+-   [Interoperability with and migration from existing C++ code](/docs/project/goals.md#interoperability-with-and-migration-from-existing-c-code)
+    -   As [future work](#function-pointers) we have the basis for a design for
+        function pointers. C++ function pointers can be modeled in Carbon code
+        as values that implement `Call`, as can member function pointers.
+
+## Alternatives considered
+
+### Signature-based function types
+
+We could give each function signature a distinct type, as is done in C and C++.
+This would provide a story for [function pointers](#function-pointers) that does
+not rely on generics or type erasure or fancy representation optimizations.
+
+The main down side of this approach is one we see frequently in C++: passing a
+function to a template generic would be less efficient than passing a function
+object to the same template. For example, in C++, given
```

Comment:
If this were solved in some other way, would that change the direction you'd want here?

I see other ways of avoiding this specific issue, so I'm curious to know how important this point of rationale ended up being.

Reply:
I think the point mentioned at the end of this section would still be relevant: giving functions unique types means that functions, lambdas, and function-like objects are more similar and have more uniform behavior. If the approach we took to avoid this specific issue affected the other cases in the same way, I think that would fully address the concern.